src/audio/smart_amp_aw/aw_smart_amp.c

```diff
@@ -0,0 +1,557 @@
+
+
+
+static int smart_amp_comp_copy(struct comp_dev *dev)
+{
+	struct smart_amp_data *smartpa_data = comp_get_drvdata(dev);
+	struct comp_buffer __sparse_cache *source_buf = buffer_acquire(smartpa_data->source_buf);
+	struct comp_buffer __sparse_cache *sink_buf = buffer_acquire(smartpa_data->sink_buf);
+	uint32_t avail_passthrough_frames;
+	uint32_t avail_feedback_frames;
+	uint32_t avail_frames;
+	uint32_t source_bytes;
+	uint32_t sink_bytes;
+	uint32_t feedback_bytes;
+
+	comp_dbg(dev, "[Awinic] smart_amp_copy()");
+
+	/* available bytes and samples calculation */
+	avail_passthrough_frames = audio_stream_avail_frames(&source_buf->stream,
+							     &sink_buf->stream);
+
+	avail_frames = avail_passthrough_frames;
+
+	if (smartpa_data->feedback_buf) {
+		struct comp_buffer __sparse_cache *feedback_buf = buffer_acquire(smartpa_data->feedback_buf);
+
+		if (comp_get_state(dev, feedback_buf->source) == dev->state) {
+			/* feedback frame calculation */
+			avail_feedback_frames =
+				audio_stream_get_avail_frames(&feedback_buf->stream);
+
+			/* RX and TX data frame alignment*/
+			avail_feedback_frames = MIN(avail_passthrough_frames,
+						    avail_feedback_frames);
+			/* calculation TX data bytes */
+			feedback_bytes = avail_feedback_frames *
+				audio_stream_frame_bytes(&feedback_buf->stream);
+
+			comp_dbg(dev, "[Awinic] smart_amp_copy(): processing %d feedback frames (avail_passthrough_frames: %d)",
+				 avail_feedback_frames, avail_passthrough_frames);
+
+			/* perform buffer writeback after source_buf process */
+			buffer_stream_invalidate(feedback_buf, feedback_bytes); 
+
+			/* copy IV data frome stream to IV buf */
+			smart_amp_fb_data_prepare(smartpa_data->algo_handle, dev, &feedback_buf->stream,
+				     avail_feedback_frames);
+
+			comp_update_buffer_consume(feedback_buf, feedback_bytes);
+		}
+
+		buffer_release(feedback_buf);
+	} else {
+		/*IV buffer clear*/
+
+	}
+
+	/* bytes calculation */
+	source_bytes = avail_frames * audio_stream_frame_bytes(&source_buf->stream);
+	sink_bytes = avail_frames * audio_stream_frame_bytes(&sink_buf->stream);
+
+	/* process data */
+	buffer_stream_invalidate(source_buf, source_bytes);
+
+    //ff data perpare
+    smart_amp_ff_data_prepare(smartpa_data->algo_handle, dev, &source_buf->stream, avail_frames);
+	
+	/*algorithm process*/
+	smart_amp_process(smartpa_data->algo_handle, dev,
+				&source_buf->stream, &sink_buf->stream, avail_frames, smartpa_data->out_channels);
+
+	buffer_stream_writeback(sink_buf, sink_bytes);
+
+	/* source/sink buffer pointers update */
+	comp_update_buffer_consume(source_buf, source_bytes);
+	comp_update_buffer_produce(sink_buf, sink_bytes);
+
+	buffer_release(sink_buf);
+	buffer_release(source_buf);
+
+	return 0;
+}
+
+static int smart_amp_comp_reset(struct comp_dev *dev)
+{
+	struct smart_amp_data *smartpa_data = comp_get_drvdata(dev);
+
+	comp_dbg(dev, "[Awinic] smart_amp_reset()");
+
+	smartpa_data->in_channels = 0;
+	smartpa_data->out_channels = 0;
```

Comment:
Does the fb buffer need to be reset even on pause? if not then the reset should go here